### DISCUSSION TITLE
Is there a way to get the document's ToC or Outline?

### DISCUSSION THREAD

**Discussion by inf3rnus** [jsvine/pdfplumber#456] [Q&A] [answered]
Hey all, sweet library!

Is it possible to get the table of contents for a pdf directly via pdfplumber?

I know it's possible via pdfminer.six via:
```
from pdfminer.pdfparser import PDFParser
from pdfminer.pdfdocument import PDFDocument

# Open a PDF document.
fp = open('mypdf.pdf', 'rb')
parser = PDFParser(fp)
document = PDFDocument(parser, password)
# Get the outlines of the document.
outlines = document.get_outlines()
```

I don't see anything in the API showing this as being possible atm unless I'm mistaken.

I'm hoping to use pdfplumber as an all in one solution and ToC's are very important for what I'm working on. I can use pdfminer.six on the side for the time being

If it's not possible at the moment - could I make a feature request?

Thanks for all your hard work :)!

Best,
Aaron

**Comment by jsvine** [comment, marked as answer]
Hi Aaron, and thanks for the thoughtful question. In `pdfplumber`, you still get (undocumented*) access to the `pdfminer.PDFDocument` object, via `.doc`. For instance:

```python
import pdfplumber
with pdfplumber.open("path/to/my.pdf") as pdf:
    print(list(pdf.doc.get_outlines()))
```

Does that work for you?

If some of the return values are `<PDFObjRef:___>` objects, you can resolve those via `pdfplumber.utils.resolve(...)` / `pdfplumber.utils.resolve_all(...)`.

\* I think it probably is worth adding more explicit documentation about this.

**Reply by inf3rnus** [reply]
That's freakin' sweet!

Thanks for the info, that covers exactly what I need, and more, now I know there's also a utils module!

Best,
Aaron

**Comment by gsgoncalves** [comment]
For others that may land here because they need the page numbers of each section.

Checkout this reference guide and sample code in the pdfminer.six repo.
https://github.com/pdfminer/pdfminer.six/blob/51683b2528e2aa685dd8b9e61f6ccf9f76a59a62/docs/source/howto/toc_target_page.rst
